add pyproject support

clarktr1 opened the pull request:
Dependabot needs a .lock file to determine dependency versions. There are other types of .lock files supported, but Pipfile.lock was the easiest to create as it supports requirements.txt. Seeing if Dependabot will the newly added files and suggestion upgrades.

Commit:
add pipfile.lock so that dependabot can detect versions

clarktr1 committed:
Update GitHub Actions runner to ubuntu-latest

Adjust action to use an up-to-date version of ubuntu

clarktr1 committed:
Update test.yml

update docker compose command

Commit:
update sphinx for tests

Commit:
update requirements

Commit:
update sphinx config

Commit:
disable pytest plugins

Commit:
update test dependencies

Commit:
update tox

Commit:
update more tox configs to allow runtests

Commit:
change tox config once again...

Commit:
change tox config once again....

Commit:
change tox config once again.....

Commit:
remove Pipfiles and update docs yaml

Commit:
adjust versions to not be as strict within range

clarktr1 commented:
@ldko @somexpert Sorry for all of the commits.

Dependabot needs a `.lock` file to determine and make version changes. I couldn't get dependabot to review the dev branch so most of the commits were me adjusting either tox configs or dependencies that clashed until all the tests passed. The current iteration passed in docker as well as through Github Action. 

Would we prefer a Pipfile.lock? Or a pyproject.toml / pyproject.lock instead?

Commit:
add pyproject&poetry with updated docker commands

Commit:
small change to pyprpoject and readme

clarktr1 commented:
@ldko @somexpert Hi! This is ready for review.

I’ve added a pyproject.toml to manage dependencies and environments and integrated [Poetry](https://python-poetry.org/). Poetry now handles dependency installation and uses poetry.lock for locking versions. Also it is Dependabot readable!

Several redundant files have been removed and their configuration consolidated into pyproject.toml:
- requirements files (base, codestyle, dev, and test)
- MANIFEST.in
- setup.py
- tox.ini

I’ve also updated Docker Compose to install and use Poetry inside the container. There should be no changes required to the existing workflow for installation or running tests.

ldko commented on README.md:
````diff
@@ -97,12 +100,4 @@ If you have SELinux, you may need to temporarily add `:Z` to the web volumes in
 To run the tests via Tox, use this command. If you are using podman-compose, swap the word `docker` with `podman` for the commands below.
 ```sh
 $ docker-compose run --rm web tox
````
The `docker-compose` commands should be changed to `docker compose` throughout the README like you did for the test.yml file.

Commit:
remove poetry from pyproject and allow  pip can handle dependencies on app installations, update README

Commit:
add lockfile

Commit:
lock poetry version, change install for docker to pip

Commit:
remove poetry from docker

Commit:
remove deprecated version header & add optional dependencies to install

Commit:
apparently version is not deprecated

Commit:
adjust tox deps groups

Commit:
remove poetry

Commit:
remove poetry lock

Commit:
remove path and curl install

Commit:
adjust contributors

clarktr1 commented:
@ldko @somexpert This is ready for a review. This MR does a few things
* Address the original dependabot issue with djangorestframework
* Add pyproject support
* Removes all requirements.txt as those are in pyproject.toml
* Update commands to use `docker compose`
* Add instructions to run this locally without Docker. 
* Update contributors


We also talked about handling vulnerabilities when we are notified by just changing the value in pyproject.toml. No lock files here!

ldko commented on README.md:
````diff
@@ -26,18 +26,20 @@ Contributors:
-```sh
-$ pip install docker-compose
-```
+There are three ways to set up this environment locally. Only use the traditional setup IF you absolutely cannot use Docker or Podman.
````
Is this "Only use the traditional setup IF you absolutely cannot use Docker or Podman." statement needed? It seems kind of strong language. I would just remove it and let people decide what they want to do for their purposes. 
```suggestion
There are three ways to set up this environment locally.
```

ldko commented on README.md:
```diff
@@ -91,18 +93,52 @@ $ sudo dnf install podman-compose
 You will follow the same steps as above, starting with `Clone the repository`. For all of the docker steps, you will have to replace the word `docker` with `podman`.
 
-If you have SELinux, you may need to temporarily add `:Z` to the web volumes in the docker-compose.yml. It will look like `.:/app/:Z`. You may also need to use `sudo` for your podman-compose commands.
+If you have SELinux, you may need to temporarily add `:Z` to the web volumes in the docker compose.yml. It will look like `.:/app/:Z`. You may also need to use `sudo` for your podman-compose commands.
```
Keep the `-` here:
```suggestion
If you have SELinux, you may need to temporarily add `:Z` to the web volumes in the docker-compose.yml. It will look like `.:/app/:Z`. You may also need to use `sudo` for your podman-compose commands.
```

ldko commented on README.md:
````diff
@@ -91,18 +93,52 @@ $ sudo dnf install podman-compose
+$ git clone https://github.com/unt-libraries/django-name.git
+$ cd django-name
 ```
 
````
I found that to run this in the "traditional setup", I need to install the MySQL development package:
```suggestion

Install the MySQL development system package. The command depends on your system, but for RedHat that is:
$  sudo dnf install mysql-devel

```

ldko commented on Dockerfile:
```diff
@@ -1,13 +1,19 @@
-RUN pip install -r requirements.txt
+
+# copy project files
+COPY pyproject.toml /app/
```
I don't think we need this line because I think the next line makes it redundant.
```suggestion
```

ldko commented on pyproject.toml:
```diff
@@ -0,0 +1,87 @@
+name = "django-name"
+description = "Name Authority App for Django."
+readme = "README.md"
+version = "1.0.0"
```
Currently we aren't on version 1.0.0 of django-name app. The setup.py previously used the `__version__` in name/__init__.py. Not sure if there is a good way to use that for project.toml, but we at least should use the version from that file here if hard coding, and if we will no longer use that variable, we should remove it from __init__.py.

clarktr1 replied:
There are some pieces within the tests, and by extension I assume the source code, that pull `__title__` and `__version__` from `__init__.py`. I think it deals with the context/brand processing. I've tried to: 

1) Just remove the values and the tests failed. 

2) Import the project metadata from`pyproject.toml` to populate the app settings at runtime. Which "worked".
- [conf.py(line 67) ](https://github.com/unt-libraries/django-name/blob/d4b4366f6575d698ffb14315b891c84b4ecbc9d0/docs/conf.py#L66) searches for the version via regex `__version__ = ___`. I _did_ change this in conf.py to just read the version metadata which worked. .But also this seems like an odd change. From what I can see, this impacts the doc's version pages.
- `pyproject` doesn't support a "Title" property and "Name" must be normalized so it would be very hacky to get the correct Title ( `django-name.replace('-', ' ').title()` ). But, looking at the app in production, it just says UNT Names. So I think this may get overridden anyways?

Here's what it currently looks like.
```python
#__init__.py
from importlib.metadata import metadata

__title__ = metadata('django-name')['Name']
__version__ = metadata('django-name')['Version']
```

Pyproject requires a version property so that can't be omitted for redundancy either. So, maybe init stays the same and we just have to track the version in both places?

ldko replied:
I wasn't concerned about the title/name since that won't regularly change; just version. Updating in both places is fine if that seems like the best option, but perhaps we add a test to ensure they are [updated together](https://packaging.python.org/en/latest/discussions/single-source-version/):

> When the intention is that a distribution package and its associated import package share the same version, it is recommended that the project include an automated test case that ensures import_name.__version__ and importlib.metadata.version("dist-name") report the same value (note: for many projects, import_name and dist-name will be the same name).

ldko commented on pyproject.toml:
```diff
@@ -0,0 +1,87 @@
+    deps = Sphinx>=4.0
+    changedir = docs
+    commands = {envbindir}/sphinx-build -W -b html . {envtmpdir}/html
+    allowlist_externals = ./runtests.py
```
Do we need this `allowlist_externals` in this section?

clarktr1 replied:
It does not. Thanks!

ldko reviewed (commented):
This is working well for me. Installed/ran locally ("traditionally") on RHEL and in Docker.

Additional notes:

1. To avoid the message: `WARN[0000] /tmp/django-name/docker-compose.yml: the attribute `version` is obsolete, it will be ignored, please remove it to avoid potential confusion`, we should delete the first line in docker-compose.yml.
2. When merging this PR, we should select `Squash and merge` from the dropdown, since there are a lot of irrelevant commits.
3. It would be helpful to rename this PR, since the current title is no longer accurate.

Commit:
add test for version checking, update readme, remove redundant steps in docker and tox

ldko commented on README.md:
```diff
@@ -93,16 +95,62 @@ You will follow the same steps as above, starting with `Clone the repository`. F
+#Ubuntu / Windows (WSL)
+sudo apt install libmysqlclient-dev
+
+#macOS via Homebrew
```
```suggestion
Install the mysql development package. The command depends on your system:
```sh
# Red Hat
sudo dnf install mysql-devel

# Ubuntu / Windows (WSL)
sudo apt install libmysqlclient-dev

# macOS via Homebrew
```

clarktr1 committed:
adjust capitalization to check git author

clarktr1 commented:
@ldko Made the suggested changes. 

1) I went ahead a made a `test_metadata.py` test suite because I didn't see any other place to put it. It's also in the style of the other test suites and does not use a class. The tox env now uses [tomli](https://pypi.org/project/tomli/) to parse pyproject, and we just use the exact same setup to read the `__version__` that is used in /docs/conf.py. 
2) Adjusted the version to 5.0.0 in `pyproject.toml`.
3) Adjusted the wording on the README.
4) Renamed the PR.
5) Removed the redundant COPY from the Dockerfile.
6) Removed `version` from `docker-compose.yml`.

ldko commented on tests/test_metadata.py:
```diff
@@ -0,0 +1,17 @@
+                                f.read(), re.MULTILINE).group(1)
+
+    pyproject_version = pyproject["project"]["version"]
+    assert pyproject_version == app_version
```
Was thinking more along the lines of what was mentioned at https://packaging.python.org/en/latest/discussions/single-source-version/, i.e.:

```suggestion
import importlib

import name


def test_versions_match():
    assert name.__version__ == importlib.metadata.version('django-name')
```

clarktr1 committed:
fix README wording and adjust version assertion

somexpert commented on .github/workflows/test.yml:
```diff
@@ -5,13 +5,13 @@ on: [push, pull_request, workflow_dispatch]
 
   tests:
-    runs-on: ubuntu-20.04
+    runs-on: ubuntu-latest
```
Should we pin this to a specific version? I know we've had issues before where eventually the version of Python we want disappears from the latest version of Ubuntu.

clarktr1 replied:
When we updated django-premis-event, django-nomination, and coda we switched from 20.04 to latest. However, aubrey-transcription and django-invite has 22.04. 

I'm not sure if there's any significance, but the former apps all are built with docker (at least in the Github tests) and the latter are not. 

I wouldn't be opposed to locking in versions for consistency.

somexpert replied:
well I guess either way we won't have a permanent solution. if we use ubuntu-latest, we have until the version of Python we want is no longer supported in the current ubuntu version, and if we pin the ubuntu version down, then we similarly will have to change things when that image gets removed from the docker hub. In either case it shouldn't be a big deal to fix, so I think you can probably just disregard this comment

somexpert replied:
I'm good with changing it or leaving it

somexpert commented on README.md:
looks like this has changed from .org to .io.
```suggestion
For installation instructions, see the [Installation](https://django-name.readthedocs.io/en/latest/installation.html) page in the docs.
```

clarktr1 commented:
@ldko Your suggestions have been addressed

ldko reviewed (approved):
Thank you, Trey! Please remember when merging this PR, to select `Squash and merge` from the dropdown, since there are a lot of irrelevant commits.

somexpert reviewed (changes requested):
Just a couple small comments above. Also let's provide a minimum for pytest-cov as we had discussed.

clarktr1 committed:
update readme with readthedocs link, add pytest-cov min version

clarktr1 commented:
@somexpert I went ahead and kept the `ubuntu-latest` for the test os. I've also changed the url to `.io` and added `https` to the hyperlink.

somexpert reviewed (approved):
This looks good to me! :thumbsup:
